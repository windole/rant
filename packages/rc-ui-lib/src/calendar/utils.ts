/* eslint-disable no-nested-ternary */
export function compareMonth(date1: Date, date2: Date): number {
  const year1 = date1.getFullYear();
  const year2 = date2.getFullYear();

  if (year1 === year2) {
    const month1 = date1.getMonth();
    const month2 = date2.getMonth();
    return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
  }

  return year1 > year2 ? 1 : -1;
}

export function compareDay(day1: Date, day2: Date): number {
  const compareMonthResult = compareMonth(day1, day2);

  if (compareMonthResult === 0) {
    const date1 = day1.getDate();
    const date2 = day2.getDate();
    return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
  }

  return compareMonthResult;
}

export const cloneDate = (date: Date): Date => new Date(date);

export const cloneDates = (dates: Date | Date[]): Date | Date[] =>
  Array.isArray(dates) ? dates.map(cloneDate) : cloneDate(dates);

export function getDayByOffset(date: Date, offset: number): Date {
  const cloned = cloneDate(date);
  cloned.setDate(cloned.getDate() + offset);
  return cloned;
}

export const getPrevDay = (date: Date): Date => getDayByOffset(date, -1);
export const getNextDay = (date: Date): Date => getDayByOffset(date, 1);
export const getToday = (): Date => {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return today;
};

export function calcDateNum(date: [Date, Date]): number {
  const day1 = date[0].getTime();
  const day2 = date[1].getTime();
  return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

const calendarLang = {
  end: '结束',
  start: '开始',
  title: '日期选择',
  confirm: '确定',
  startEnd: '开始/结束',
  weekdays: ['日', '一', '二', '三', '四', '五', '六'],
  monthTitle: (year: number, month: number) => `${year}年${month}月`,
  rangePrompt: (maxRange: number) => `最多选择 ${maxRange} 天`,
};

export const t = (key: string, ...args): string => {
  if (args.length) {
    return calendarLang[key](...args);
  }
  return calendarLang[key];
};

export const formatMonthTitle = (date: Date): string =>
  t('monthTitle', date.getFullYear(), date.getMonth() + 1);
