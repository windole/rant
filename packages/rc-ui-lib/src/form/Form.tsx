import React, { forwardRef, useContext } from 'react';
import classnames from 'classnames';
import RcForm from 'rc-field-form';
import type { FormInstance } from 'rc-field-form';
import { FormProps } from './PropsType';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import { FormContext } from './FormContext';
import { CellGroup } from '../cell';

const Form = forwardRef<FormInstance, FormProps>((props, ref) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('form', prefixCls);

  const {
    className,
    style,
    layout,
    footer,
    children,
    border,
    inset,
    colon,
    showValidateMessage,
    ...formProps
  } = props;

  return (
    <FormContext.Provider
      value={{
        layout,
        colon,
        showValidateMessage,
        validateTrigger: formProps.validateTrigger,
      }}
    >
      <RcForm className={classnames(bem(), className)} style={style} ref={ref} {...formProps}>
        <CellGroup border={border} inset={inset}>
          {children}
        </CellGroup>
        {footer && <div className={classnames(bem('footer'))}>{footer}</div>}
      </RcForm>
    </FormContext.Provider>
  );
});

Form.defaultProps = {
  showValidateMessage: true,
  validateTrigger: 'onChange',
};

export default Form;
