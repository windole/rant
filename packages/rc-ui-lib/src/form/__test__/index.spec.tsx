import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';
import { sleep } from '../../../tests/utils';
import { toArray } from '../../uploader/utils';
import Button from '../../button';
import Field from '../../field';
import Form from '..';

export type InternalNamePath = (string | number)[];
export type NamePath = string | number | InternalNamePath;
const changeValue = async (wrapper, value) => {
  wrapper.find('input').simulate('change', { target: { value } });
  await act(async () => {
    await sleep();
  });
  wrapper.update();
};

export function getNamePath(path: NamePath | null): InternalNamePath {
  return toArray(path);
}

export function matchNamePath(
  namePath: InternalNamePath,
  changedNamePath: InternalNamePath | null,
) {
  if (!namePath || !changedNamePath || namePath.length !== changedNamePath.length) {
    return false;
  }
  return namePath.every((nameUnit, i) => changedNamePath[i] === nameUnit);
}
export function getField(wrapper, index: string | number = 0) {
  if (typeof index === 'number') {
    return wrapper.find(Field).at(index);
  }

  const name = getNamePath(index);
  const fields = wrapper.find(Field);
  for (let i = 0; i < fields.length; i += 1) {
    const field = fields.at(i);
    const fieldName = getNamePath(field.props().name);

    if (matchNamePath(name, fieldName)) {
      return field;
    }
  }
  return null;
}

describe('Form', () => {
  let wrapper;
  let spyConsole: jest.SpyInstance;
  beforeEach(() => {
    spyConsole = jest.spyOn(console, 'warn');
    spyConsole.mockImplementation(() => {
      return null;
    });
  });

  afterEach(() => {
    wrapper.unmount();
    spyConsole.mockRestore();
    jest.restoreAllMocks();
  });

  it('basic usage', async () => {
    const onClick = jest.fn();
    wrapper = mount(
      <Form
        layout="horizontal"
        colon
        inset
        footer={
          <div style={{ margin: '16px 16px 0' }}>
            <Button round nativeType="submit" type="primary" block>
              提交
            </Button>
          </div>
        }
      >
        <Form.Item
          disabled
          tooltip={{
            message:
              'A prime is a natural number greater than 1 that has no positive divisors other than 1 and itself.',
          }}
          intro="确保这是唯一的用户名"
          rules={[{ required: true, message: '请填写用户名' }]}
          name="username"
          label="用户名"
        >
          <Field placeholder="请输入用户名" />
        </Form.Item>
        <Form.Item
          rules={[{ required: false, message: '请填写密码' }]}
          label="密码"
          className="my-password"
          colon
          labelWidth="100"
          labelAlign="center"
          labelClass="my-label-class"
          onClick={onClick}
        >
          <Field placeholder="请输入密码" />
        </Form.Item>
      </Form>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('when noStyle props is set', async () => {
    wrapper = mount(
      <Form>
        <Form.Item noStyle>
          <Field placeholder="请输入密码" />
        </Form.Item>
      </Form>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('when shouldUpdate props is set', async () => {
    wrapper = mount(
      <Form>
        <Form.Item shouldUpdate>{() => <Field placeholder="请输入密码" />}</Form.Item>
      </Form>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("shouldn't work when shouldUpdate is set", async () => {
    const spy = jest.fn();
    wrapper = mount(
      <Form>
        <Form.Item dependencies={['field_2']} shouldUpdate={() => true}>
          {() => {
            spy();
            return 'gogogo';
          }}
        </Form.Item>
        <Form.Item name="field_1">
          <Field />
        </Form.Item>
        <Form.Item name="field_2">
          <Field />
        </Form.Item>
      </Form>,
    );

    expect(spy).toHaveBeenCalledTimes(1);
    await changeValue(getField(wrapper, 1), 'value1');
    // sync start
    //   valueUpdate -> rerender by shouldUpdate
    //   depsUpdate  -> rerender by deps
    //   [ react rerender once -> 2 ]
    // sync end
    expect(spy).toHaveBeenCalledTimes(2);

    await changeValue(getField(wrapper, 2), 'value2');
    // sync start
    //   valueUpdate -> rerender by shouldUpdate
    //   depsUpdate  -> rerender by deps
    //   [ react rerender once -> 3 ]
    // sync end
    expect(spy).toHaveBeenCalledTimes(3);

    expect(spyConsole).toHaveBeenLastCalledWith(
      "[FormItem] `shouldUpdate` and `dependencies` shouldn't be used together.",
    );
  });

  it('when rules without reqiured props is set', async () => {
    wrapper = mount(
      <Form>
        <Form.Item rules={[{ pattern: /\d{6}/, message: '请输入6位数字' }]}>
          <Field placeholder="请输入密码" />
        </Form.Item>
      </Form>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('when rules with reqiured props is set', async () => {
    wrapper = mount(
      <Form>
        <Form.Item rules={[{ pattern: /\d{6}/, message: '请输入6位数字', required: true }]}>
          <Field placeholder="请输入密码" />
        </Form.Item>
      </Form>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('when name props is set without shouldUpdate or dependencies', async () => {
    wrapper = mount(
      <Form>
        <Form.Item name="name">{() => <Field placeholder="请输入密码" />}</Form.Item>
      </Form>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(spyConsole).toHaveBeenCalledTimes(2);
  });

  it('when shouldUpdate and dependencies are not set', async () => {
    wrapper = mount(
      <Form>
        <Form.Item>{() => <Field placeholder="请输入密码" />}</Form.Item>
      </Form>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(spyConsole).toHaveBeenLastCalledWith(
      '[FormItem] `children` of render props only work with `shouldUpdate` or `dependencies`.',
    );
  });

  it('dependencies is set and name is not set ', async () => {
    let x = 0;
    wrapper = mount(
      <Form>
        <Form.Item dependencies={['field_1']}>
          {() => {
            x += 1;
            return `gogogo${x}`;
          }}
        </Form.Item>
        <Form.Item name="field_1">
          <input />
        </Form.Item>
      </Form>,
    );
  });

  it('when children is not validElement', async () => {
    const handleBlur = jest.fn();
    wrapper = mount(
      <Form>
        <Form.Item name="test" trigger="blur">
          <Field placeholder="请输入密码" onBlur={handleBlur} />
        </Form.Item>
      </Form>,
    );
    wrapper.find('input').simulate('focus');
    await sleep(20);
    await wrapper.find('input').simulate('blur');
    expect(handleBlur).toHaveBeenCalled();
  });
});
