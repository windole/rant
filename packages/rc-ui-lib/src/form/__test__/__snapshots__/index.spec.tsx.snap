// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form basic usage 1`] = `
<ForwardRef
  colon={true}
  footer={
    <div
      style={
        Object {
          "margin": "16px 16px 0",
        }
      }
    >
      <Button
        block={true}
        iconPosition="left"
        nativeType="submit"
        round={true}
        size="normal"
        tag="button"
        type="primary"
      >
        提交
      </Button>
    </div>
  }
  inset={true}
  layout="horizontal"
  showValidateMessage={true}
  validateTrigger="onChange"
>
  <ForwardRef(Form2)
    className="rc-form"
    validateTrigger="onChange"
  >
    <form
      className="rc-form"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <CellGroup
        border={true}
        inset={true}
      >
        <div
          className="rc-cell-group rc-cell-group--inset"
        >
          <FormItem
            disabled={true}
            intro="确保这是唯一的用户名"
            label="用户名"
            name="username"
            rules={
              Array [
                Object {
                  "message": "请填写用户名",
                  "required": true,
                },
              ]
            }
            tooltip={
              Object {
                "message": "A prime is a natural number greater than 1 that has no positive divisors other than 1 and itself.",
              }
            }
          >
            <WrapperField
              name="username"
              rules={
                Array [
                  Object {
                    "message": "请填写用户名",
                    "required": true,
                  },
                ]
              }
              validateTrigger="onChange"
            >
              <Field2
                fieldContext={
                  Object {
                    "_init": true,
                    "getFieldError": [Function],
                    "getFieldValue": [Function],
                    "getFieldWarning": [Function],
                    "getFieldsError": [Function],
                    "getFieldsValue": [Function],
                    "getInternalHooks": [Function],
                    "isFieldTouched": [Function],
                    "isFieldValidating": [Function],
                    "isFieldsTouched": [Function],
                    "isFieldsValidating": [Function],
                    "resetFields": [Function],
                    "setFieldValue": [Function],
                    "setFields": [Function],
                    "setFieldsValue": [Function],
                    "submit": [Function],
                    "validateFields": [Function],
                    "validateTrigger": "onChange",
                  }
                }
                isListField={false}
                key="_username"
                name={
                  Array [
                    "username",
                  ]
                }
                rules={
                  Array [
                    Object {
                      "message": "请填写用户名",
                      "required": true,
                    },
                  ]
                }
                trigger="onChange"
                validateTrigger="onChange"
                valuePropName="value"
              >
                <FormItemLayout
                  disabled={true}
                  htmlFor="username"
                  intro="确保这是唯一的用户名"
                  isFieldChildren={true}
                  label="用户名"
                  meta={
                    Object {
                      "errors": Array [],
                      "name": Array [
                        "username",
                      ],
                      "touched": false,
                      "validated": false,
                      "validating": false,
                      "warnings": Array [],
                    }
                  }
                  required={true}
                  tooltip={
                    Object {
                      "message": "A prime is a natural number greater than 1 that has no positive divisors other than 1 and itself.",
                    }
                  }
                >
                  <ForwardRef
                    className="rc-form"
                    clearIcon="clear"
                    clearTrigger="focus"
                    colon={true}
                    disabled={true}
                    error={false}
                    errorMessage={null}
                    formatTrigger="onChange"
                    htmlFor="username"
                    id="username"
                    intro="确保这是唯一的用户名"
                    label="用户名"
                    onChange={[Function]}
                    placeholder="请输入用户名"
                    required={true}
                    tooltip={
                      Object {
                        "message": "A prime is a natural number greater than 1 that has no positive divisors other than 1 and itself.",
                      }
                    }
                    validateTrigger="onChange"
                    value=""
                  >
                    <Cell
                      className="rc-field rc-field--disabled rc-form"
                      icon={null}
                      required={true}
                      title={
                        <React.Fragment>
                          用户名
                          :
                          <div
                            className="rc-field__tooltip"
                            onClick={[Function]}
                          >
                            <Icon
                              name="question-o"
                              tag="i"
                            />
                          </div>
                        </React.Fragment>
                      }
                      titleClass="rc-field__label"
                      titleStyle={Object {}}
                      valueClass="rc-field__value"
                    >
                      <div
                        className="rc-cell rc-cell--required rc-field rc-field--disabled rc-form"
                      >
                        <div
                          className="rc-cell__title rc-field__label"
                          style={Object {}}
                        >
                          用户名
                          :
                          <div
                            className="rc-field__tooltip"
                            onClick={[Function]}
                          >
                            <Icon
                              name="question-o"
                              tag="i"
                            >
                              <Badge
                                className="van-icon van-icon-question-o"
                                showZero={true}
                                style={
                                  Object {
                                    "color": undefined,
                                    "fontSize": undefined,
                                  }
                                }
                                tag="i"
                              >
                                <i
                                  className="rc-badge__wrapper van-icon van-icon-question-o"
                                  style={
                                    Object {
                                      "color": undefined,
                                      "fontSize": undefined,
                                    }
                                  }
                                />
                              </Badge>
                            </Icon>
                          </div>
                        </div>
                        <div
                          className="rc-cell__value rc-field__value"
                        >
                          <div
                            className="rc-field__body"
                          >
                            <input
                              className="rc-field__control"
                              disabled={true}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onCompositionEnd={[Function]}
                              onCompositionStart={[Function]}
                              onCompositionUpdate={[Function]}
                              onFocus={[Function]}
                              onKeyPress={[Function]}
                              placeholder="请输入用户名"
                              value=""
                            />
                          </div>
                          <div
                            className="rc-field__intro"
                          >
                            确保这是唯一的用户名
                          </div>
                        </div>
                      </div>
                    </Cell>
                  </ForwardRef>
                </FormItemLayout>
              </Field2>
            </WrapperField>
          </FormItem>
          <FormItem
            className="my-password"
            colon={true}
            label="密码"
            labelAlign="center"
            labelClass="my-label-class"
            labelWidth="100"
            onClick={[MockFunction]}
            rules={
              Array [
                Object {
                  "message": "请填写密码",
                  "required": false,
                },
              ]
            }
          >
            <FormItemLayout
              className="my-password"
              colon={true}
              isFieldChildren={true}
              label="密码"
              labelAlign="center"
              labelClass="my-label-class"
              labelWidth="100"
              onClick={[MockFunction]}
            >
              <ForwardRef
                className="rc-form my-password"
                clearIcon="clear"
                clearTrigger="focus"
                colon={true}
                errorMessage={null}
                formatTrigger="onChange"
                label="密码"
                labelAlign="center"
                labelClass="my-label-class"
                labelWidth="100"
                onClick={[MockFunction]}
                placeholder="请输入密码"
                validateTrigger="onChange"
              >
                <Cell
                  className="rc-field rc-field--label-center rc-form my-password"
                  icon={null}
                  onClick={[MockFunction]}
                  title={
                    <React.Fragment>
                      密码
                      :
                    </React.Fragment>
                  }
                  titleClass="rc-field__label rc-field__label--center my-label-class"
                  titleStyle={
                    Object {
                      "width": "100px",
                    }
                  }
                  valueClass="rc-field__value"
                >
                  <div
                    className="rc-cell rc-field rc-field--label-center rc-form my-password"
                    onClick={[MockFunction]}
                  >
                    <div
                      className="rc-cell__title rc-field__label rc-field__label--center my-label-class"
                      style={
                        Object {
                          "width": "100px",
                        }
                      }
                    >
                      密码
                      :
                    </div>
                    <div
                      className="rc-cell__value rc-field__value"
                    >
                      <div
                        className="rc-field__body"
                      >
                        <input
                          className="rc-field__control"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onCompositionEnd={[Function]}
                          onCompositionStart={[Function]}
                          onCompositionUpdate={[Function]}
                          onFocus={[Function]}
                          onKeyPress={[Function]}
                          placeholder="请输入密码"
                        />
                      </div>
                    </div>
                  </div>
                </Cell>
              </ForwardRef>
            </FormItemLayout>
          </FormItem>
        </div>
      </CellGroup>
      <div
        className="rc-form__footer"
      >
        <div
          style={
            Object {
              "margin": "16px 16px 0",
            }
          }
        >
          <Button
            block={true}
            iconPosition="left"
            nativeType="submit"
            round={true}
            size="normal"
            tag="button"
            type="primary"
          >
            <button
              className="rc-button rc-button--primary rc-button--normal rc-button--block rc-button--round"
              onClick={[Function]}
              style={Object {}}
              type="submit"
            >
              <div
                className="rc-button__content"
              >
                <span
                  className="rc-button__text"
                  key="text"
                >
                  提交
                </span>
              </div>
            </button>
          </Button>
        </div>
      </div>
    </form>
  </ForwardRef(Form2)>
</ForwardRef>
`;

exports[`Form when name props is set without shouldUpdate or dependencies 1`] = `
<ForwardRef
  showValidateMessage={true}
  validateTrigger="onChange"
>
  <ForwardRef(Form2)
    className="rc-form"
    validateTrigger="onChange"
  >
    <form
      className="rc-form"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <CellGroup
        border={true}
      >
        <div
          className="rc-cell-group rc-hairline--top-bottom"
        >
          <FormItem
            name="name"
          >
            <WrapperField
              name="name"
              validateTrigger="onChange"
            >
              <Field2
                fieldContext={
                  Object {
                    "_init": true,
                    "getFieldError": [Function],
                    "getFieldValue": [Function],
                    "getFieldWarning": [Function],
                    "getFieldsError": [Function],
                    "getFieldsValue": [Function],
                    "getInternalHooks": [Function],
                    "isFieldTouched": [Function],
                    "isFieldValidating": [Function],
                    "isFieldsTouched": [Function],
                    "isFieldsValidating": [Function],
                    "resetFields": [Function],
                    "setFieldValue": [Function],
                    "setFields": [Function],
                    "setFieldsValue": [Function],
                    "submit": [Function],
                    "validateFields": [Function],
                    "validateTrigger": "onChange",
                  }
                }
                isListField={false}
                key="_name"
                name={
                  Array [
                    "name",
                  ]
                }
                trigger="onChange"
                validateTrigger="onChange"
                valuePropName="value"
              >
                <FormItemLayout
                  htmlFor="name"
                  meta={
                    Object {
                      "errors": Array [],
                      "name": Array [
                        "name",
                      ],
                      "touched": false,
                      "validated": false,
                      "validating": false,
                      "warnings": Array [],
                    }
                  }
                  required={false}
                >
                  <ForwardRef
                    className="rc-form"
                    clearIcon="clear"
                    clearTrigger="focus"
                    error={false}
                    errorMessage={null}
                    formatTrigger="onChange"
                    htmlFor="name"
                    required={false}
                    validateTrigger="onChange"
                  >
                    <Cell
                      className="rc-field rc-form"
                      icon={null}
                      required={false}
                      title={null}
                      titleClass="rc-field__label"
                      titleStyle={Object {}}
                      valueClass="rc-field__value"
                    >
                      <div
                        className="rc-cell rc-field rc-form"
                      >
                        <div
                          className="rc-cell__value rc-cell__value--alone rc-field__value"
                        >
                          <div
                            className="rc-field__body"
                          >
                            <input
                              className="rc-field__control"
                              onBlur={[Function]}
                              onChange={[Function]}
                              onCompositionEnd={[Function]}
                              onCompositionStart={[Function]}
                              onCompositionUpdate={[Function]}
                              onFocus={[Function]}
                              onKeyPress={[Function]}
                              placeholder=""
                            />
                          </div>
                        </div>
                      </div>
                    </Cell>
                  </ForwardRef>
                </FormItemLayout>
              </Field2>
            </WrapperField>
          </FormItem>
        </div>
      </CellGroup>
    </form>
  </ForwardRef(Form2)>
</ForwardRef>
`;

exports[`Form when noStyle props is set 1`] = `
<ForwardRef
  showValidateMessage={true}
  validateTrigger="onChange"
>
  <ForwardRef(Form2)
    className="rc-form"
    validateTrigger="onChange"
  >
    <form
      className="rc-form"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <CellGroup
        border={true}
      >
        <div
          className="rc-cell-group rc-hairline--top-bottom"
        >
          <FormItem
            noStyle={true}
          >
            <ForwardRef
              clearIcon="clear"
              clearTrigger="focus"
              formatTrigger="onChange"
              placeholder="请输入密码"
            >
              <Cell
                className="rc-field"
                icon={null}
                title={null}
                titleClass="rc-field__label"
                titleStyle={Object {}}
                valueClass="rc-field__value"
              >
                <div
                  className="rc-cell rc-field"
                >
                  <div
                    className="rc-cell__value rc-cell__value--alone rc-field__value"
                  >
                    <div
                      className="rc-field__body"
                    >
                      <input
                        className="rc-field__control"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onCompositionEnd={[Function]}
                        onCompositionStart={[Function]}
                        onCompositionUpdate={[Function]}
                        onFocus={[Function]}
                        onKeyPress={[Function]}
                        placeholder="请输入密码"
                      />
                    </div>
                  </div>
                </div>
              </Cell>
            </ForwardRef>
          </FormItem>
        </div>
      </CellGroup>
    </form>
  </ForwardRef(Form2)>
</ForwardRef>
`;

exports[`Form when rules with reqiured props is set 1`] = `
<ForwardRef
  showValidateMessage={true}
  validateTrigger="onChange"
>
  <ForwardRef(Form2)
    className="rc-form"
    validateTrigger="onChange"
  >
    <form
      className="rc-form"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <CellGroup
        border={true}
      >
        <div
          className="rc-cell-group rc-hairline--top-bottom"
        >
          <FormItem
            rules={
              Array [
                Object {
                  "message": "请输入6位数字",
                  "pattern": /\\\\d\\{6\\}/,
                  "required": true,
                },
              ]
            }
          >
            <FormItemLayout
              isFieldChildren={true}
            >
              <ForwardRef
                className="rc-form"
                clearIcon="clear"
                clearTrigger="focus"
                errorMessage={null}
                formatTrigger="onChange"
                placeholder="请输入密码"
                validateTrigger="onChange"
              >
                <Cell
                  className="rc-field rc-form"
                  icon={null}
                  title={null}
                  titleClass="rc-field__label"
                  titleStyle={Object {}}
                  valueClass="rc-field__value"
                >
                  <div
                    className="rc-cell rc-field rc-form"
                  >
                    <div
                      className="rc-cell__value rc-cell__value--alone rc-field__value"
                    >
                      <div
                        className="rc-field__body"
                      >
                        <input
                          className="rc-field__control"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onCompositionEnd={[Function]}
                          onCompositionStart={[Function]}
                          onCompositionUpdate={[Function]}
                          onFocus={[Function]}
                          onKeyPress={[Function]}
                          placeholder="请输入密码"
                        />
                      </div>
                    </div>
                  </div>
                </Cell>
              </ForwardRef>
            </FormItemLayout>
          </FormItem>
        </div>
      </CellGroup>
    </form>
  </ForwardRef(Form2)>
</ForwardRef>
`;

exports[`Form when rules without reqiured props is set 1`] = `
<ForwardRef
  showValidateMessage={true}
  validateTrigger="onChange"
>
  <ForwardRef(Form2)
    className="rc-form"
    validateTrigger="onChange"
  >
    <form
      className="rc-form"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <CellGroup
        border={true}
      >
        <div
          className="rc-cell-group rc-hairline--top-bottom"
        >
          <FormItem
            rules={
              Array [
                Object {
                  "message": "请输入6位数字",
                  "pattern": /\\\\d\\{6\\}/,
                },
              ]
            }
          >
            <FormItemLayout
              isFieldChildren={true}
            >
              <ForwardRef
                className="rc-form"
                clearIcon="clear"
                clearTrigger="focus"
                errorMessage={null}
                formatTrigger="onChange"
                placeholder="请输入密码"
                validateTrigger="onChange"
              >
                <Cell
                  className="rc-field rc-form"
                  icon={null}
                  title={null}
                  titleClass="rc-field__label"
                  titleStyle={Object {}}
                  valueClass="rc-field__value"
                >
                  <div
                    className="rc-cell rc-field rc-form"
                  >
                    <div
                      className="rc-cell__value rc-cell__value--alone rc-field__value"
                    >
                      <div
                        className="rc-field__body"
                      >
                        <input
                          className="rc-field__control"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onCompositionEnd={[Function]}
                          onCompositionStart={[Function]}
                          onCompositionUpdate={[Function]}
                          onFocus={[Function]}
                          onKeyPress={[Function]}
                          placeholder="请输入密码"
                        />
                      </div>
                    </div>
                  </div>
                </Cell>
              </ForwardRef>
            </FormItemLayout>
          </FormItem>
        </div>
      </CellGroup>
    </form>
  </ForwardRef(Form2)>
</ForwardRef>
`;

exports[`Form when shouldUpdate and dependencies are not set 1`] = `
<ForwardRef
  showValidateMessage={true}
  validateTrigger="onChange"
>
  <ForwardRef(Form2)
    className="rc-form"
    validateTrigger="onChange"
  >
    <form
      className="rc-form"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <CellGroup
        border={true}
      >
        <div
          className="rc-cell-group rc-hairline--top-bottom"
        >
          <FormItem>
            <WrapperField
              validateTrigger="onChange"
            >
              <Field2
                fieldContext={
                  Object {
                    "_init": true,
                    "getFieldError": [Function],
                    "getFieldValue": [Function],
                    "getFieldWarning": [Function],
                    "getFieldsError": [Function],
                    "getFieldsValue": [Function],
                    "getInternalHooks": [Function],
                    "isFieldTouched": [Function],
                    "isFieldValidating": [Function],
                    "isFieldsTouched": [Function],
                    "isFieldsValidating": [Function],
                    "resetFields": [Function],
                    "setFieldValue": [Function],
                    "setFields": [Function],
                    "setFieldsValue": [Function],
                    "submit": [Function],
                    "validateFields": [Function],
                    "validateTrigger": "onChange",
                  }
                }
                isListField={false}
                key="_"
                trigger="onChange"
                validateTrigger="onChange"
                valuePropName="value"
              >
                <FormItemLayout
                  htmlFor=""
                  meta={
                    Object {
                      "errors": Array [],
                      "name": Array [],
                      "touched": false,
                      "validated": false,
                      "validating": false,
                      "warnings": Array [],
                    }
                  }
                  required={false}
                >
                  <ForwardRef
                    className="rc-form"
                    clearIcon="clear"
                    clearTrigger="focus"
                    error={false}
                    errorMessage={null}
                    formatTrigger="onChange"
                    htmlFor=""
                    required={false}
                    validateTrigger="onChange"
                  >
                    <Cell
                      className="rc-field rc-form"
                      icon={null}
                      required={false}
                      title={null}
                      titleClass="rc-field__label"
                      titleStyle={Object {}}
                      valueClass="rc-field__value"
                    >
                      <div
                        className="rc-cell rc-field rc-form"
                      >
                        <div
                          className="rc-cell__value rc-cell__value--alone rc-field__value"
                        >
                          <div
                            className="rc-field__body"
                          >
                            <input
                              className="rc-field__control"
                              onBlur={[Function]}
                              onChange={[Function]}
                              onCompositionEnd={[Function]}
                              onCompositionStart={[Function]}
                              onCompositionUpdate={[Function]}
                              onFocus={[Function]}
                              onKeyPress={[Function]}
                              placeholder=""
                            />
                          </div>
                        </div>
                      </div>
                    </Cell>
                  </ForwardRef>
                </FormItemLayout>
              </Field2>
            </WrapperField>
          </FormItem>
        </div>
      </CellGroup>
    </form>
  </ForwardRef(Form2)>
</ForwardRef>
`;

exports[`Form when shouldUpdate props is set 1`] = `
<ForwardRef
  showValidateMessage={true}
  validateTrigger="onChange"
>
  <ForwardRef(Form2)
    className="rc-form"
    validateTrigger="onChange"
  >
    <form
      className="rc-form"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <CellGroup
        border={true}
      >
        <div
          className="rc-cell-group rc-hairline--top-bottom"
        >
          <FormItem
            shouldUpdate={true}
          >
            <WrapperField
              shouldUpdate={true}
              validateTrigger="onChange"
            >
              <Field2
                fieldContext={
                  Object {
                    "_init": true,
                    "getFieldError": [Function],
                    "getFieldValue": [Function],
                    "getFieldWarning": [Function],
                    "getFieldsError": [Function],
                    "getFieldsValue": [Function],
                    "getInternalHooks": [Function],
                    "isFieldTouched": [Function],
                    "isFieldValidating": [Function],
                    "isFieldsTouched": [Function],
                    "isFieldsValidating": [Function],
                    "resetFields": [Function],
                    "setFieldValue": [Function],
                    "setFields": [Function],
                    "setFieldsValue": [Function],
                    "submit": [Function],
                    "validateFields": [Function],
                    "validateTrigger": "onChange",
                  }
                }
                isListField={false}
                key="_"
                shouldUpdate={true}
                trigger="onChange"
                validateTrigger="onChange"
                valuePropName="value"
              >
                <FormItemLayout
                  htmlFor=""
                  meta={
                    Object {
                      "errors": Array [],
                      "name": Array [],
                      "touched": false,
                      "validated": false,
                      "validating": false,
                      "warnings": Array [],
                    }
                  }
                  required={false}
                >
                  <ForwardRef
                    className="rc-form"
                    clearIcon="clear"
                    clearTrigger="focus"
                    error={false}
                    errorMessage={null}
                    formatTrigger="onChange"
                    htmlFor=""
                    required={false}
                    validateTrigger="onChange"
                  >
                    <Cell
                      className="rc-field rc-form"
                      icon={null}
                      required={false}
                      title={null}
                      titleClass="rc-field__label"
                      titleStyle={Object {}}
                      valueClass="rc-field__value"
                    >
                      <div
                        className="rc-cell rc-field rc-form"
                      >
                        <div
                          className="rc-cell__value rc-cell__value--alone rc-field__value"
                        >
                          <div
                            className="rc-field__body"
                          >
                            <div
                              className="rc-field__control rc-field__control--custom"
                            >
                              <ForwardRef
                                clearIcon="clear"
                                clearTrigger="focus"
                                formatTrigger="onChange"
                                placeholder="请输入密码"
                              >
                                <Cell
                                  className="rc-field"
                                  icon={null}
                                  title={null}
                                  titleClass="rc-field__label"
                                  titleStyle={Object {}}
                                  valueClass="rc-field__value"
                                >
                                  <div
                                    className="rc-cell rc-field"
                                  >
                                    <div
                                      className="rc-cell__value rc-cell__value--alone rc-field__value"
                                    >
                                      <div
                                        className="rc-field__body"
                                      >
                                        <input
                                          className="rc-field__control"
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          onCompositionEnd={[Function]}
                                          onCompositionStart={[Function]}
                                          onCompositionUpdate={[Function]}
                                          onFocus={[Function]}
                                          onKeyPress={[Function]}
                                          placeholder="请输入密码"
                                        />
                                      </div>
                                    </div>
                                  </div>
                                </Cell>
                              </ForwardRef>
                            </div>
                          </div>
                        </div>
                      </div>
                    </Cell>
                  </ForwardRef>
                </FormItemLayout>
              </Field2>
            </WrapperField>
          </FormItem>
        </div>
      </CellGroup>
    </form>
  </ForwardRef(Form2)>
</ForwardRef>
`;
